File: week-4/permutations/permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = [];
    const visited = {};
    recurse(nums, 0, [], visited, res);
    return res;
};

const recurse = (nums, index, cur_array, visited, res) => {
    if (cur_array.length === nums.length) {
        res.push(cur_array.slice());
        return;
    }
    
    for (let i=0; i<nums.length; i++) {
        if (nums[i] in visited && visited[nums[i]]) continue;
        cur_array.push(nums[i]);
        visited[nums[i]] = true;
        recurse(nums, i, cur_array, visited, res);
        cur_array.pop();
        visited[nums[i]] = false;
    }
};

File: week-4/permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        visited = {}
        self.recurse(nums, 0, [], visited,res)
        return res
    
    def recurse(self, nums, index, cur_array, visited, res):
        if len(cur_array) == len(nums):
            res.append(cur_array[:])
            return
        
        for i in range(0, len(nums)):
            if nums[i] in visited and visited[nums[i]]: continue
            cur_array.append(nums[i])
            visited[nums[i]] = True
            self.recurse(nums, i, cur_array, visited, res)
            cur_array.pop()
            visited[nums[i]] = False
